crates/arroyo-connectors/src/iggy/mod.rs

```diff
@@ -0,0 +1,624 @@
+                done: true,
+                message: message.message,
+            });
+        });
+
+        Some(rx)
+    }
+
+    fn test(
+        &self,
+        _: &str,
+        config: Self::ProfileT,
+        table: Self::TableT,
+        schema: Option<&ConnectionSchema>,
+        tx: Sender<TestSourceMessage>,
+    ) {
+        let tester = IggyTester { connection: config };
+        tester.start(table, schema.cloned(), tx);
+    }
+
+    fn table_type(&self, _: Self::ProfileT, table: Self::TableT) -> ConnectionType {
+        match table.type_ {
+            TableType::Source { .. } => ConnectionType::Source,
+            TableType::Sink { .. } => ConnectionType::Sink,
+        }
+    }
+
+    fn from_options(
+        &self,
+        name: &str,
+        options: &mut arroyo_rpc::ConnectorOptions,
+        schema: Option<&ConnectionSchema>,
+        _profile: Option<&ConnectionProfile>,
+    ) -> Result<ArroyoConnection> {
+        let endpoint = options.pull_str("endpoint")?;
+        let transport = options.pull_str("transport")?;
+
+        let authentication = if let Some(username) = options.pull_opt_str("username")? {
+            let password = options.pull_str("password")?;
+            IggyConfigAuthentication::UsernamePassword {
+                username,
+                password: VarStr::new(password),
+            }
+        } else {
+            IggyConfigAuthentication::None {}
+        };
+
+        let transport_protocol = match transport.as_str() {
+            "http" => TransportProtocol::Http,
+            _ => bail!("Invalid transport protocol"),
+        };
+
+        let config = IggyConfig {
+            endpoint: endpoint.clone(),
+            transport: transport_protocol,
+            authentication: authentication.clone(),
+        };
+
+        let stream = options.pull_str("stream")?;
+        let topic = options.pull_str("topic")?;
+
+        let table_type = if options.pull_bool("is_source")? {
+            let offset = match options.pull_str("offset")?.as_str() {
+                "earliest" => "earliest",
+                "latest" => "latest",
+                _ => bail!("Invalid offset value"),
+            };
+
+            let consumer_id = options.pull_u64("consumer_id")? as u32;
+            let partition_id = options.pull_u64("partition_id")? as u32;
+            let auto_commit = options.pull_bool("auto_commit")?;
+
+            TableType::Source {
+                offset: match offset {
+                    "earliest" => SourceOffset::Earliest,
+                    "latest" => SourceOffset::Latest,
+                    _ => bail!("Invalid offset value"),
+                },
+                consumer_id: consumer_id as i64,
+                partition_id: partition_id as i64,
+                auto_commit,
+            }
+        } else {
+            let partitioning = match options.pull_str("partitioning")?.as_str() {
+                "partition_id" => "partition_id",
```

Comment:
Similarly here, we're first validating the strings and then on line 190 converting to enums, we can just do that once